Fix MRP baseTimeout calculation for LIT

yunhanw-google opened the pull request:
Currently MRP baseTimeout calculation may be fine for SIT, but it is problematic for LIT ICD where Idle internal could be very large, and mrp timeout could be very large, which prevent devices going into sleep node.

Actually, we need to handle two scenarios: the controller sending an initial message to a LIT, and the controller sending a subsequent message. 

This PR matched the logic from GetRetransmissionTimeout with ReliableMessageMgr::CalculateNextRetransTime which always treats things as "active" for non-initial messages on the exchange.

Fixes: https://github.com/project-chip/connectedhomeip/issues/35354

#### Testing
Working on manually triggering and testing this case locally with manufactuers

yunhanw-google committed:
Fix MRP baseTimeout calculation when it is possible the other end fall asleep

Damian-Nordic commented on src/messaging/ReliableMessageProtocolConfig.cpp:
```diff
@@ -135,7 +135,7 @@ System::Clock::Timeout GetRetransmissionTimeout(System::Clock::Timeout activeInt
     System::Clock::Timestamp timeout(0);
     for (uint8_t i = 0; i < CHIP_CONFIG_RMP_DEFAULT_MAX_RETRANS + 1; i++)
     {
-        auto baseInterval = ((timeSinceLastActivity + timeout) < activityThreshold) ? activeInterval : idleInterval;
```
You shouldn't remove the usage of idle interval for SIT or non-ICD devices. This exception should be for LIT only where the idle interval can be very long.

yunhanw-google replied:
I agree, I think it would be to use min(activeInterval, idleInterval) here for both client and server regarding mrp configuration.

For SIT/non-ICD, SAI would be larger than SII, then it would go with SII when it reaches to idle status.

For LIT, SAI would be less than SII, then it would go with SAI when it reaches to idle status.

Damian-Nordic replied:
I don't think SAI is usually larger than SII for SIT/non-ICD. In fact, default SAI is 300ms while SII is 500. For SIT it's also common to use a sub-second fast-polling interval and a couple of seconds slow-polling interval. Unfortunately, it seems that we need some sort of flag passed from outside to indicate that the receiver is LIT. Alternatively, we could detect that by checking if SII > 15s?

yunhanw-google replied:
Yes, considering dynamic ICD feature, it would be hard to feature flag to setup MRP timeout here.

I would prefer checking if SII > 15 s, I just update the code to reflect, thanks

yunhanw-google committed:
use min(SAI, SII) for mrp timeout when clock runs into idle

restyled-commits committed:
Restyled by clang-format

bzbarsky-apple commented on src/messaging/ReliableMessageProtocolConfig.cpp:
```diff
@@ -135,7 +135,8 @@ System::Clock::Timeout GetRetransmissionTimeout(System::Clock::Timeout activeInt
     {
-        auto baseInterval = ((timeSinceLastActivity + timeout) < activityThreshold) ? activeInterval : idleInterval;
+        auto baseInterval =
+            ((timeSinceLastActivity + timeout) < activityThreshold) ? activeInterval : min(activeInterval, idleInterval);
```
This is a clear spec violation.

yunhanw-google replied:
I just follow @damian's suggestion, and use 15 seconds to determine whether device is in LIT mode, then if
yes, set the timeout base as SAI when timeSinceLastActivity + timeout) >= activityThreshold and it is LIT

bzbarsky-apple replied:
> I just follow @damian's suggestion in this pr

Again: it's as spec violation.

> For the original https://github.com/project-chip/connectedhomeip/issues/35354#issuecomment-2342005193, i don't think that is enough

So:

1. The changes proposed in this bug are not enough to fix the fact that the timeout calculations are just wrong.
2. I don't understand your example of "if client send out the first case message in the end of active window".  The issue with timeouts we are seeing is that a server sending a Sigma2 waits long enough that a client sending it a message and using its SII would time out MRP before it times out waiting for the Sigma2 response.  But of course no client is going to use SII in that calculation of actual MRP bits, since it's an active exchange.  So properly using SAI in that situation will in fact fix https://github.com/project-chip/connectedhomeip/issues/35354.

> but session layer may not be able to know whether it is the first message..

Of course, which is why the information about whether things should be force-considered not idle needs to be passed in to the session from the API consumer, which _does_ have this information about the state of the exchange.

yunhanw-google replied:
Thanks for comments. I agree. 

I have matched the logic from GetRetransmissionTimeout the logic with ReliableMessageMgr::CalculateNextRetransTime which always treats things as "active" for non-initial messages on the exchange. 

i have plumber the isPeerActive boolean from application to session layer. The client, like commissioner/controller, would always use SAI for calculation via setting isPeerActive as true.

yunhanw-google replied:
updated, thanks

bzbarsky-apple commented on src/transport/SecureSession.h:
```diff
@@ -265,7 +265,8 @@ class SecureSession : public Session, public ReferenceCounted<SecureSession, Sec
     {
-        return IsPeerActive() ? GetRemoteMRPConfig().mActiveRetransTimeout : GetRemoteMRPConfig().mIdleRetransTimeout;
+        return IsPeerActive() ? GetRemoteMRPConfig().mActiveRetransTimeout
+                              : min(GetRemoteMRPConfig().mActiveRetransTimeout, GetRemoteMRPConfig().mIdleRetransTimeout);
```
This is also a spec violation.

yunhanw-google replied:
reverted, thanks

bzbarsky-apple commented on src/transport/UnauthenticatedSessionTable.h:
```diff
@@ -155,7 +155,8 @@ class UnauthenticatedSession : public Session, public ReferenceCounted<Unauthent
     {
-        return IsPeerActive() ? GetRemoteMRPConfig().mActiveRetransTimeout : GetRemoteMRPConfig().mIdleRetransTimeout;
+        return IsPeerActive() ? GetRemoteMRPConfig().mActiveRetransTimeout
+                              : min(GetRemoteMRPConfig().mActiveRetransTimeout, GetRemoteMRPConfig().mIdleRetransTimeout);
```
As is this.

yunhanw-google replied:
reverted, thanks

bzbarsky-apple reviewed (changes requested):
The right fix for https://github.com/project-chip/connectedhomeip/issues/35354 is to fix the issue pointed out in https://github.com/project-chip/connectedhomeip/issues/35354#issuecomment-2342005193

yunhanw-google committed:
update mrp calculation

-- Use 15 seconds to determine whether device is in LIT mode, then if
yes, set the timeout base as SAI if GetRemoteMRPConfig().mIdleRetransTimeout.count() > sitIcdSlowPollMaximum.count()

yunhanw-google committed:
Address comments

bzbarsky-apple commented on src/app/CommandSender.cpp:
```diff
@@ -125,7 +125,8 @@ CHIP_ERROR CommandSender::SendCommandRequestInternal(const SessionHandle & sessi
 
-    mExchangeCtx->SetResponseTimeout(timeout.ValueOr(session->ComputeRoundTripTimeout(app::kExpectedIMProcessingTime)));
+    mExchangeCtx->SetResponseTimeout(timeout.ValueOr(
+        session->ComputeRoundTripTimeout(app::kExpectedIMProcessingTime, mExchangeCtx.Get()->HasReceivedAtLeastOneMessage())));
```
Won't `mExchangeCtx.Get()->HasReceivedAtLeastOneMessage()` always be false here?  I don't get this logic.

yunhanw-google replied:
removed. thanks

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -254,7 +254,7 @@ class Session
     // side to receive our message (accounting for our MRP retransmits if it
     // gets lost) and send a response.
-    virtual System::Clock::Milliseconds32 GetAckTimeout() const = 0;
+    virtual System::Clock::Milliseconds32 GetAckTimeout(bool isPeerActive) const = 0;
```
What does this boolean really mean?  Is it a "we know the peer is active via some information not available to the session"?  Or is it a "the MRP machinery on our side will assume the peer is active regardless of what the session state is"?

I suspect it's the latter of those two, but please double-check.  The naming, documentation (which needs to exist) and actual setting of this parameter need to match the intended meaning.

yunhanw-google replied:
yes, the latter should be right after checking around the code. updated. thanks

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -265,15 +265,16 @@ class Session
     // indicate "peer is reaching out to us, not in response to anything").
-    virtual System::Clock::Milliseconds32 GetMessageReceiptTimeout(System::Clock::Timestamp ourLastActivity) const = 0;
+    virtual System::Clock::Milliseconds32 GetMessageReceiptTimeout(System::Clock::Timestamp ourLastActivity,
+                                                                   bool isPeerActive) const = 0;
```
This part makes no sense to me.  Whether the _peer_ is active or not has nothing to do with how long it will keep retrying to get a message to _us_.  What matter is whether the peer thinks _we_ are active.

In practice we need to carefully look at the callsites of this function.  I suspect in all of them the peer would be assuming we are active because it's responding to our message, but that needs to be verified carefully.

yunhanw-google replied:
I agree, yes, after checking all calls, i think all of them the peer would be assuming we are active. updated, thanks

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -265,15 +265,16 @@ class Session
     // characteristics of the target and a caller-provided value for the time it takes to process a message at the upper layer on
     // the target For group sessions, this function will always return 0.
-    System::Clock::Timeout ComputeRoundTripTimeout(System::Clock::Timeout upperlayerProcessingTimeout);
+    System::Clock::Timeout ComputeRoundTripTimeout(System::Clock::Timeout upperlayerProcessingTimeout, bool isPeerActive);
```
Again, what does this boolean actually mean?

yunhanw-google replied:
removed, thanks

bzbarsky-apple commented on src/messaging/ReliableMessageProtocolConfig.cpp:
```diff
@@ -126,7 +126,8 @@ Optional<ReliableMessageProtocolConfig> GetLocalMRPConfig()
 System::Clock::Timeout GetRetransmissionTimeout(System::Clock::Timeout activeInterval, System::Clock::Timeout idleInterval,
-                                                System::Clock::Timeout lastActivityTime, System::Clock::Timeout activityThreshold)
+                                                System::Clock::Timeout lastActivityTime, System::Clock::Timeout activityThreshold,
+                                                bool isPeerActive)
```
So... maybe if we want to match `ReliableMessageMgr::CalculateNextRetransTime` we should just explicitly have a isResponseMessage boolean argument here?  That would make it _very_ clear what is actually going on and how to go about deciding what to pass in.

yunhanw-google replied:
I am now using isInitial boolean, which should be similar to !isResponseMessage, thanks

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.h:
```diff
@@ -157,13 +157,16 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
 
     FabricIndex GetFabricIndex() const { return mFabricIndex; }
 
+    bool IsPeerActive() { return mExchangeCtxt.HasValue() && mExchangeCtxt.Value()->HasReceivedAtLeastOneMessage(); };
```
Should not be needed.

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.h:
```diff
@@ -157,13 +157,16 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
     // Compute our Sigma1 response timeout.  This can give consumers an idea of
     // how long it will take to detect that our Sigma1 did not get through.
-    static System::Clock::Timeout ComputeSigma1ResponseTimeout(const ReliableMessageProtocolConfig & remoteMrpConfig);
+    static System::Clock::Timeout ComputeSigma1ResponseTimeout(const ReliableMessageProtocolConfig & remoteMrpConfig,
+                                                               bool isPeerActive);
```
For a _Sigma1_, again if we are trying to match CalculateNextRetransTime(), the peer is never "known active", right?  Because it's the first message on the exchange.

yunhanw-google replied:
i agree, updated, thanks

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.h:
```diff
@@ -157,13 +157,16 @@ class DLL_EXPORT CASESession : public Messaging::UnsolicitedMessageHandler,
     // Compute our Sigma2 response timeout.  This can give consumers an idea of
     // how long it will take to detect that our Sigma1 did not get through.
-    static System::Clock::Timeout ComputeSigma2ResponseTimeout(const ReliableMessageProtocolConfig & remoteMrpConfig);
+    static System::Clock::Timeout ComputeSigma2ResponseTimeout(const ReliableMessageProtocolConfig & remoteMrpConfig,
```
Sigma2 is always a response.  So in this case the peer is always known "active".

yunhanw-google replied:
i agree, updated, thanks

yunhanw-google committed:
address comments

restyled-commits committed:
Restyled by whitespace

yunhanw-google commented on src/transport/Session.h:
```diff
@@ -265,7 +265,9 @@ class Session
-    virtual System::Clock::Milliseconds32 GetMessageReceiptTimeout(System::Clock::Timestamp ourLastActivity) const = 0;
+    // isInitial indicates whether the other side treats us as initial message.
+    virtual System::Clock::Milliseconds32 GetMessageReceiptTimeout(System::Clock::Timestamp ourLastActivity,
+                                                                   bool isInitial) const = 0;
```
I agree, thanks, updated

yunhanw-google committed:
address comments

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -2622,23 +2622,26 @@ System::Clock::Timeout ComputeRoundTripTimeout(ExchangeContext::Timeout serverPr
+                                    System::Clock::kZero, remoteMrpConfig.mActiveThresholdTime, isFirstMessageOnExchange) +
         serverProcessingTime +
         GetRetransmissionTimeout(localMRPConfig.mActiveRetransTimeout, localMRPConfig.mIdleRetransTimeout,
                                  // Peer will assume we are active, since it's
```
I can't suggest on the whole comment here, since it includes removals, but this comment should say:
```
    // Peer will be responding to our message, so isFirstMessageOnExchange should be false
    // and the timestamp does not matter.
```

or something along those lines.

yunhanw-google replied:
updated, thanks

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -2622,23 +2622,26 @@ System::Clock::Timeout ComputeRoundTripTimeout(ExchangeContext::Timeout serverPr
 System::Clock::Timeout CASESession::ComputeSigma1ResponseTimeout(const ReliableMessageProtocolConfig & remoteMrpConfig)
 {
-    return ComputeRoundTripTimeout(kExpectedSigma1ProcessingTime, remoteMrpConfig);
+    // Assume it is the initial message on the session, and treat peer as idle for Sigma1
+    return ComputeRoundTripTimeout(kExpectedSigma1ProcessingTime, remoteMrpConfig, false /*isFirstMessageOnExchange*/);
```
```suggestion
    // Sigma1 is the first message on the exchange.
    return ComputeRoundTripTimeout(kExpectedSigma1ProcessingTime, remoteMrpConfig, true /*isFirstMessageOnExchange*/);
```

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -2622,23 +2622,26 @@ System::Clock::Timeout ComputeRoundTripTimeout(ExchangeContext::Timeout serverPr
 System::Clock::Timeout CASESession::ComputeSigma2ResponseTimeout(const ReliableMessageProtocolConfig & remoteMrpConfig)
 {
-    return ComputeRoundTripTimeout(kExpectedHighProcessingTime, remoteMrpConfig);
+    // Assume it is the response message on the session, and treat peer as active for Sigma2
+    return ComputeRoundTripTimeout(kExpectedHighProcessingTime, remoteMrpConfig, true /*isFirstMessageOnExchange*/);
```
```suggestion
    // Sigma2 is never the first message on the exchange.
    return ComputeRoundTripTimeout(kExpectedHighProcessingTime, remoteMrpConfig, false /*isFirstMessageOnExchange*/);
```

yunhanw-google replied:
updated, thanks

bzbarsky-apple commented on src/transport/SecureSession.h:
```diff
@@ -194,7 +195,8 @@ class SecureSession : public Session, public ReferenceCounted<SecureSession, Sec
             return GetRetransmissionTimeout(localMRPConfig.mActiveRetransTimeout, localMRPConfig.mIdleRetransTimeout,
-                                            ourLastActivity, localMRPConfig.mActiveThresholdTime);
+                                            ourLastActivity, localMRPConfig.mActiveThresholdTime,
+                                            ourLastActivity == System::Clock::kZero /*isFirstMessageOnExchange*/);
```
```suggestion
                                            false /*isFirstMessageOnExchange*/);
```
`GetMessageReceiptTimeout` is always asking how long the _response_ will take, so it's never the first message on an exchange, no?

bzbarsky-apple commented on src/transport/Session.cpp:
```diff
@@ -71,9 +71,10 @@ System::Clock::Timeout Session::ComputeRoundTripTimeout(System::Clock::Timeout u
-    // other side would be responding to our message.
-    return GetAckTimeout() + upperlayerProcessingTimeout + GetMessageReceiptTimeout(System::SystemClock().GetMonotonicTimestamp());
+    // Treat us as active for purposes of GetMessageReceiptTimeout(), pass false into GetAckTimeout to
+    // indicate we are processing non-initial message since the other side would be responding to our message.
```
No, the "responding to our message" is the `GetMessageReceiptTimeout()` bit, and it passes false already.  But GetAckTimeout.... we don't know what to pass it, because it depends on whether this ComputeRoundTripTimeout() call is for an initial message or not.  As in, this function needs a isFirstMessageOnExchange argument, just like the copy of this function in CASESession.cpp needed one.

yunhanw-google replied:
audited and update all ComputeRoundTripTimeout caller, thanks

bzbarsky-apple commented on src/transport/Session.h:
```diff
@@ -254,7 +254,8 @@ class Session
     // side to receive our message (accounting for our MRP retransmits if it
     // gets lost) and send a response.
-    virtual System::Clock::Milliseconds32 GetAckTimeout() const = 0;
+    // This need to know whether we are talking about an initial message we send or not via isFirstMessageOnExchange
```
```suggestion
    //
    // The MRP retransmit delay computation depends on whether this is the first message on the exchange.
```

bzbarsky-apple commented on src/transport/UnauthenticatedSessionTable.h:
```diff
@@ -120,7 +121,8 @@ class UnauthenticatedSession : public Session, public ReferenceCounted<Unauthent
             return GetRetransmissionTimeout(localMRPConfig.mActiveRetransTimeout, localMRPConfig.mIdleRetransTimeout,
-                                            ourLastActivity, localMRPConfig.mActiveThresholdTime);
+                                            ourLastActivity, localMRPConfig.mActiveThresholdTime,
+                                            ourLastActivity == System::Clock::kZero /*isFirstMessageOnExchange*/);
```
```suggestion
                                            false /*isFirstMessageOnExchange*/);
```

yunhanw-google committed:
Update src/transport/UnauthenticatedSessionTable.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/transport/Session.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/protocols/secure_channel/CASESession.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/transport/SecureSession.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/protocols/secure_channel/CASESession.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/app/ReadClient.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/messaging/ReliableMessageProtocolConfig.cpp

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
Update src/messaging/ReliableMessageProtocolConfig.h

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

yunhanw-google committed:
address comments

yunhanw-google committed:
address comments

yunhanw-google committed:
update true isFirstMessageOnExchange for all ComputeRoundTripTimeout caller

restyled-commits committed:
Restyled by whitespace

restyled-commits committed:
Restyled by clang-format

yunhanw-google committed:
fix compilation error

yunhanw-google committed:
address comments

yunhanw-google committed:
cleanup

bzbarsky-apple commented on src/messaging/ExchangeContext.h:
```diff
@@ -177,7 +177,6 @@ class DLL_EXPORT ExchangeContext : public ReliableMessageContext,
 
     // Applies a suggested response timeout value based on the session type and the given upper layer processing time for
     // the next message to the exchange. The exchange context must have a valid session when calling this function.
-    //
```
Why this change?  Please don't make this change.

yunhanw-google replied:
done, thanks

bzbarsky-apple commented on src/protocols/secure_channel/CASESession.cpp:
```diff
@@ -546,6 +546,7 @@ CHIP_ERROR CASESession::EstablishSession(SessionManager & sessionManager, Fabric
     mSessionResumptionStorage = sessionResumptionStorage;
     mLocalMRPConfig           = MakeOptional(mrpLocalConfig.ValueOr(GetDefaultMRPConfig()));
 
+    // Sigma1 is the first message on the exchange.
```
```suggestion
```

Damian-Nordic commented on src/transport/SecureSession.h:
```diff
@@ -194,7 +195,8 @@ class SecureSession : public Session, public ReferenceCounted<SecureSession, Sec
             return GetRetransmissionTimeout(localMRPConfig.mActiveRetransTimeout, localMRPConfig.mIdleRetransTimeout,
-                                            ourLastActivity, localMRPConfig.mActiveThresholdTime);
+                                            ourLastActivity, localMRPConfig.mActiveThresholdTime,
+                                            false /*isFirstMessageOnExchange*/);
```
Shouldn't this also be taken as a parameter? This is used by `ReadClient` to calculate the timeout for receiving ReportData and each ReportData comes on a new exchange, so in such a case this parameter should be `true`, right?

yunhanw-google replied:
see this comment from boris, 
https://github.com/project-chip/connectedhomeip/pull/37696#discussion_r1984363287

where "GetMessageReceiptTimeout is always asking how long the response will take, so it's never the first message on an exchange."
See comment here. https://github.com/project-chip/connectedhomeip/blob/714907ec5b146dbf48373b72a32fd838e26402a5/src/app/ReadClient.cpp#L991

https://github.com/project-chip/connectedhomeip/blob/714907ec5b146dbf48373b72a32fd838e26402a5/src/transport/Session.cpp#L76

Damian-Nordic replied:
But notice that if you pass `false` here then passing `kZero` in `ReadCient` does nothing, so this comment ("Peer will assume we are idle (hence we pass kZero to GetMessageReceiptTimeout())") becomes irrelevant. That is, we will always calculate the receipt time as if the peer thought we were active.

yunhanw-google replied:
yes, it is, I would prefer follow-up comment improvement on this part, thanks

bzbarsky-apple replied:
Hmm.  So the API documentation for GetMessageReceiptTimeout says:

```
    // GetReceiptTimeout is the estimate for how long it could take for us to
    // receive a message after the other side sends it, accounting for the MRP
    // retransmits the other side might do if the message gets lost.
```

which I thought would mean it's always a response but of course @Damian-Nordic is right and ReadClient is using it for the initial message from the other side.  So this PR can't go in as-is, because it would in fact mis-compute things there and then subscriptions would not work right.

Maybe the right thing is for GetMessageReceiptTimeout to lose the timestamp bit entirely, because the only values we pass are zero to indicate "initial message" and "now" to indicate "a response", and instead have an actual `isFirstMessageOnExchange` boolean arg on GetMessageReceiptTimeout?  This would be false in ReadClient, true in the one other caller in Session.cpp...  I guess `GetRetransmissionTimeout` still takes a "last activity" thing, but that could be computed in the implementation of GetMessageReceiptTimeout: zero if isFirstMessageOnExchange, now otherwise?

bzbarsky-apple replied:
Alternately, we could in fact add a boolean arg to GetMessageReceiptTimeout and just pass it through, after fixing the two callsites, and then do a followup to figure out whether we can fix GetRetransmissionTimeout to not require the activity time or something...

yunhanw-google replied:
updated to add a boolean arg to GetMessageReceiptTimeout,  I think we need set isFirstMessageOnExchange as **true** in readClient, set isFirstMessageOnExchange as **false** in one 

> instead have an actual isFirstMessageOnExchange boolean arg on GetMessageReceiptTimeout? This would be **false** in ReadClient, **true** in the one other caller in Session.cpp

Damian-Nordic replied:
Thanks for fixing this. I agree with Boris that perhaps the timestamp argument in GetMessageReceiptTimeout is not really needed anymore, but it can be removed in a followup as well.

bzbarsky-apple replied:
> I think we need set isFirstMessageOnExchange as true in readClient, set isFirstMessageOnExchange as false in session side

Yes, sorry, I had it backwards above......

bzbarsky-apple commented on src/transport/Session.cpp:
```diff
@@ -71,10 +71,10 @@ System::Clock::Timeout Session::ComputeRoundTripTimeout(System::Clock::Timeout u
-    // Treat us as active for purposes of GetMessageReceiptTimeout(), pass false into GetAckTimeout to
+    // Treat us as active for purposes of GetMessageReceiptTimeout(), pass false into GetAckTimeout and GetMessageReceiptTimeout to
     // indicate we are processing non-initial message since the other side would be responding to our message.
     return GetAckTimeout(false /*isFirstMessageOnExchange*/) + upperlayerProcessingTimeout +
```
So hold on.  Now that I am re-reading this... (and I am sorry we keep going back and forth):

1. GetAckTimeout is how long we expect the message from us to them to take, right?
2. GetMessageReceiptTimeout is how long we expect the message from them to us to take.

For the second one, isFirstMessageOnExchange should definitely be false, since it's responding to our message.

But for the first (GetAckTimeout....  isFirstMessageOnExchange needs to be set based on whether we are computing the roundtrip bits for the first message on the exchange or not, no?

I thought this had gotten fixed in https://github.com/project-chip/connectedhomeip/pull/37696#discussion_r1985465438 so what exactly is going on here?

bzbarsky-apple replied:
> [force-pushed](https://github.com/project-chip/connectedhomeip/compare/c3a4999a123991f547e8bd9bfa08817fc0d833c5..e3fad9b13b5c62f0fa73b014341f2b1810d932a7) the feature/case_lit_timeout branch from [c3a4999](https://github.com/project-chip/connectedhomeip/commit/c3a4999a123991f547e8bd9bfa08817fc0d833c5) to [e3fad9b](https://github.com/project-chip/connectedhomeip/commit/e3fad9b13b5c62f0fa73b014341f2b1810d932a7)

This is where the broken came in.  See https://github.com/project-chip/connectedhomeip/blob/c3a4999/src/transport/Session.h#L279-L280 which has the right signature, but then it went away.

yunhanw-google replied:
i use vscode github UI to checkout the multiple PRs, somehow it create a new branch automatically with same PRs during PR switches..and lose the right change sets....

working on the right branch

yunhanw-google replied:
I locate the right branch, and compared it against the last comments you make when you mark the previous approval, and add the latest change that fix GetMessageReceiptTimeou

yunhanw-google committed:
add parameter isFirstMessageOnExchange in GetMessageReceiptTimeout

restyled-commits committed:
Restyled by clang-format
